### PR TITLE
FMDN upstream part 1: Crypto

### PR DESCRIPTION
**Pull request by kapi-no**

### CODE REVIEW

**Review by MarekPieta — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto [hunk 1]
```diff
@@ -64,6 +64,16 @@ config BT_FAST_PAIR_CRYPTO_PSA
 
 endchoice
 
+# Define Fast Pair operations that can be supported by each backend.
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: These are not optional for the FMDN extension. They are required. When we reviewed this part, we aimed to introduce similar options for other configurations. If we do not support the PN extension some APIs and ops are also not needed

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c [hunk 2]
```diff
@@ -61,3 +94,56 @@ int fp_crypto_ecdh_shared_secret(uint8_t *secret_key,
...
+				      uint8_t *mod,
+				      const uint8_t *in,
+				      size_t datalen)
+{
+	 /* SECP160R1 key length corresponds to one coordinate. */
+	uint8_t public_key[FP_CRYPTO_ECC_SECP160R1_KEY_LEN * 2];
+	ocrypto_sc_p160 mod_le;
+
+	/* Verify the input length. */
+	if (datalen != SECP160R1_DATA_LEN) {
```
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:108]: It's a limitation of this particular backend. See the mbedTLS implementation

[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:108]: The SECP functions are very similar to existing FP crypto APIs that have similar usage (e.g. hashing function):

```
int fp_crypto_sha256(uint8_t *out, const uint8_t *in, size_t data_len);
```

File: subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: I mean that they are optional from `fp_crypto` backend's perspective (backend can declare if it implements given crypto operation or not). Still, looking at your response, my comment by itself may be misleading.
Maybe, apart from comment update, we could mark crypto operations required by FMDN under Kconfig helps?

[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: Keep in mind that FMDN is an user of `fp_crypto` library, so the `fp_crypto` library does not need to know about FMDN requirements (FMDN tailors `fp_crypto` to align it to FMDN needs)

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:108]: I meant that, if FMDN uses only a single input length here, we could simplify API. Still, I would not insist and we can keep it as is if you prefer more generic implementation.

**Review by alstrzebonski — APPROVED** [review]

File: subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: > Keep in mind that FMDN is an user of `fp_crypto` library, so the `fp_crypto` library does not need to know about FMDN requirements (FMDN tailors `fp_crypto` to align it to FMDN needs)

The same thing can be said about the PN extension and its dedicated crypto requirements and the `fp_crypto_additional_data_encode` function

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c [hunk 3]
```diff
@@ -61,3 +94,56 @@ int fp_crypto_ecdh_shared_secret(uint8_t *secret_key,
 
 	return 0;
 }
+
+int fp_crypto_ecc_secp160r1_calculate(uint8_t *out,
+				      uint8_t *mod,
+				      const uint8_t *in,
+				      size_t datalen)
+{
+	 /* SECP160R1 key length corresponds to one coordinate. */
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c]: I think compliance generally does not complain about spaces after tabs (it seems not to have special handling for comments)

File: subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto (hunk 1 above)
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: I would say it's unlikely that these FMDN crypto operations would require other data length. For additional data encode chances that various data length are needed are higher I think (e.g. if a new operation is added). Still, we can keep the API as is here.

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c [hunk 4]
```diff
@@ -9,11 +9,30 @@
...
+
+#include <zephyr/sys/byteorder.h>
 
 #include <zephyr/logging/log.h>
 LOG_MODULE_DECLARE(fp_crypto, CONFIG_FP_CRYPTO_LOG_LEVEL);
 
+#define SECP160R1_DATA_LEN (32U)
+#define SECP256R1_DATA_LEN (32U)
+
+/* START: importing additional APIs from the Oberon runtime. */
```
[the remark this answers is not in the document]
Comment (MarekPieta) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:25]: @kapi-no, maybe we can open a tracking ticket and get back to it after this PR is merged?

File: subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto (hunk 1 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/Kconfig.fp_crypto:70]: Discussed offline:

```suggestion
# A backend supporting a given crypto operation selects a related Kconfig option.
```

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:25]: discussed offline, create a tracking ticket

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:108]: discussed offline, keep as is

File: subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (kapi-no) [reply subsys/bluetooth/services/fast_pair/fp_crypto/fp_crypto_oberon.c:25]: Defined a tracking ticket in JIRA:

https://nordicsemi.atlassian.net/browse/NCSDK-27410

Please do not resolve this comment thread so it is easier to access it using a link defined in the JIRA ticket

**Comment by kapi-no** [comment]
The last force push contains only the rebase diff

**Review by MarekPieta — APPROVED** [review]

**Comment by kapi-no** [comment]
Dropping the dependency on nrfxlib as it was merged into the main NCS branch.

**Pull request merged**